### PR TITLE
eksctl install flux now handles re-installation (by deleting only Flux-related resources, and re-creating them)

### LINKED ISSUE #1084: Handle "flux already installed in cluster" situation
If flux is already installed in the cluster eksctl should handle it earlier.

```
$ EKSCTL_EXPERIMENTAL=true eksctl install flux -f examples/01-simple-cluster.yaml --git-url=git@github.com:martina-if/flux-test-2.git --git-email=martina+flux@weave.works
[ℹ]  Cloning git@github.com:martina-if/flux-test-2.git
Cloning into '/tmp/eksctl-install-flux-clone194001168'...
warning: templates not found /home/martina/.git_template
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (3/3), done.
[ℹ]  Writing Flux manifests
[ℹ]  Installing Flux into the cluster
[ℹ]  replaced "flux:Deployment.apps/flux"
[ℹ]  replaced "flux:Secret/flux-git-deploy"
[ℹ]  replaced "flux:Deployment.apps/memcached"
[✖]  You may find the local clone of git@github.com:martina-if/flux-test-2.git used by eksctl at /tmp/eksctl-install-flux-clone194001168
[✖]  Service "memcached" is invalid: spec.clusterIP: Invalid value: "": field is immutable
```

```
$ eksctl version
[ℹ]  version.Info{BuiltAt:"1564746220", GitCommit:"0.2.0-46-g35bedce0", GitTag:""}
```

### PR DESCRIPTION
**Pull request by marccarre**
### Description

`eksctl install flux` now handles re-installation by deleting all resources and creating them again.

Other approaches were considered, like simply using `kubectl apply`
but given this would add an external dependency, this solution was
chosen for now.

Also note that the `Secret` for Flux's private SSH key isn't changed, in order to avoid the user having to reconfigure Flux access to their repository. (This can be verified in the below manual test by the fact the public SSH key given to the user is the same.)

Fixes #1084. Follows-up on #1203 (never merged).

### Checklist

- [x] Code compiles correctly (i.e `make build`)
- [x] ~Added tests that cover your change (if possible)~ I didn't add any E2E test for this given the cost associated to them, and given this is an edge case.
- [x] All unit tests passing (i.e. `make test`)
- [x] ~Added/modified documentation as required (such as the `README.md`, and `examples` directory)~
- [x] Manually tested

### Manual test

```console
$ EKSCTL_EXPERIMENTAL=true AWS_PROFILE=default-mfa \
    ./eksctl \
        -f examples/01-simple-cluster.yaml \
        install flux \
        --git-url=git@github.com:marccarre/my-gitops-repo.git \
        --git-email=carre.marc@gmail.com

[ℹ]  Generating public key infrastructure for the Helm Operator and Tiller
[ℹ]    this may take up to a minute, please be patient
[!]  Public key infrastructure files were written into directory "/var/folders/24/d3mml6bn20nftpt91cfldq1h0000gn/T/eksctl-helm-pki727902620"
[!]  please move the files into a safe place or delete them
[ℹ]  Generating manifests
[ℹ]  Cloning git@github.com:marccarre/my-gitops-repo.git
Cloning into '/var/folders/24/d3mml6bn20nftpt91cfldq1h0000gn/T/eksctl-install-flux-clone-387791435'...
remote: Enumerating objects: 48, done.
remote: Counting objects: 100% (48/48), done.
remote: Compressing objects: 100% (37/37), done.
remote: Total 190 (delta 11), reused 47 (delta 11), pack-reused 142
Receiving objects: 100% (190/190), 81.24 KiB | 388.00 KiB/s, done.
Resolving deltas: 100% (51/51), done.
[ℹ]  Writing Flux manifests
[ℹ]  Applying Helm TLS Secret(s)
[ℹ]  created "Namespace/flux"
[ℹ]  created "flux:Secret/flux-helm-tls-cert"
[ℹ]  created "flux:Secret/tiller-secret"
[!]  Note: certificate secrets aren't added to the Git repository for security reasons
[ℹ]  Applying manifests
[ℹ]  replaced "Namespace/flux"
[ℹ]  created "flux:ConfigMap/flux-helm-tls-ca-config"
[ℹ]  created "flux:ServiceAccount/flux"
[ℹ]  created "ClusterRole.rbac.authorization.k8s.io/flux"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/flux"
[ℹ]  created "flux:Deployment.apps/flux"
[ℹ]  created "flux:ServiceAccount/tiller"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/tiller"
[ℹ]  created "flux:ServiceAccount/helm"
[ℹ]  created "flux:Role.rbac.authorization.k8s.io/tiller-user"
[ℹ]  created "kube-system:RoleBinding.rbac.authorization.k8s.io/tiller-user-binding"
[ℹ]  created "flux:Deployment.apps/memcached"
[ℹ]  created "flux:ServiceAccount/flux-helm-operator"
[ℹ]  created "ClusterRole.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  created "flux:Service/memcached"
[ℹ]  created "flux:Secret/flux-git-deploy"
[ℹ]  created "flux:Deployment.extensions/tiller-deploy"
[ℹ]  created "flux:Service/tiller-deploy"
[ℹ]  created "CustomResourceDefinition.apiextensions.k8s.io/helmreleases.helm.fluxcd.io"
[ℹ]  created "flux:Deployment.apps/flux-helm-operator"
[ℹ]  Waiting for Helm Operator to start
ERROR: logging before flag.Parse: E0829 17:40:47.084976   94575 portforward.go:331] an error occurred forwarding 49776 -> 3030: error forwarding port 3030 to pod 128bb0f4414ec9483ed29dad9bf3c483cee3d418e0052080a8d7221aa37d52cd, uid : exit status 1: 2019/08/29 08:40:47 socat[20733] E connect(5, AF=2 127.0.0.1:3030, 16): Connection refused
[!]  Helm Operator is not ready yet (Get http://127.0.0.1:49776/healthz: EOF), retrying ...
[...]
[ℹ]  Helm Operator started successfully
[ℹ]  see https://docs.fluxcd.io/projects/helm-operator for details on how to use the Helm Operator
[ℹ]  Waiting for Flux to start
[ℹ]  Flux started successfully
[ℹ]  see https://docs.fluxcd.io/projects/flux for details on how to use Flux
[ℹ]  Committing and pushing manifests to git@github.com:marccarre/my-gitops-repo.git
[master a14f424] Add Initial Flux configuration
 1 file changed, 38 insertions(+), 38 deletions(-)
 rewrite flux/tiller-ca-cert-configmap.yaml (87%)
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 8 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 2.15 KiB | 2.15 MiB/s, done.
Total 4 (delta 1), reused 0 (delta 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To github.com:marccarre/my-gitops-repo.git
   6631168..a14f424  master -> master
[ℹ]  Flux will only operate properly once it has write-access to the Git repository
[ℹ]  please configure git@github.com:marccarre/my-gitops-repo.git so that the following Flux SSH public key has write access to it
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC6wyI1Zzc7bFJgpWbrTA6xISdPeavT9zLdbMGf/X9NBmJt+wj1TG0SOGkMSpS670KoNdWu8KdSh77uPkif9MDn2f/oaOdfMwq6nOAQoPdTHpmdx6friliLkoV0RozFkuLPWfg3xXQIkPanRX6APUivOXoqI4jRXt9Ghj62BZkYDNIzzkQ5dOwbZB+nv4+q9X8fC4okMlU68QWosan4qb1mDB0LwIYyhqTAqwTd2aWww+/e0jy/vEdWdD1LVTVb4S8DDcXiijFa32nbVmU40Xm5xBH9+1jNLo/REM7PjuVvah1XFmKpyPZhYglPe/bJsHH+A9P2pY9OEsq/4dzCTQFx
```
```console
$ EKSCTL_EXPERIMENTAL=true AWS_PROFILE=default-mfa \
    ./eksctl \
        -f examples/01-simple-cluster.yaml \
        install flux \
        --git-url=git@github.com:marccarre/my-gitops-repo.git \
        --git-email=carre.marc@gmail.com

[ℹ]  Generating public key infrastructure for the Helm Operator and Tiller
[ℹ]    this may take up to a minute, please be patient
[!]  Public key infrastructure files were written into directory "/var/folders/24/d3mml6bn20nftpt91cfldq1h0000gn/T/eksctl-helm-pki491471785"
[!]  please move the files into a safe place or delete them
[ℹ]  Generating manifests
[ℹ]  Cloning git@github.com:marccarre/my-gitops-repo.git
Cloning into '/var/folders/24/d3mml6bn20nftpt91cfldq1h0000gn/T/eksctl-install-flux-clone-370994932'...
remote: Enumerating objects: 52, done.
remote: Counting objects: 100% (52/52), done.
remote: Compressing objects: 100% (40/40), done.
remote: Total 194 (delta 12), reused 51 (delta 12), pack-reused 142
Receiving objects: 100% (194/194), 83.34 KiB | 287.00 KiB/s, done.
Resolving deltas: 100% (52/52), done.
[ℹ]  Writing Flux manifests
[ℹ]  Applying Helm TLS Secret(s)
[ℹ]  deleted "flux:Secret/flux-helm-tls-cert"
[ℹ]  deleted "flux:Secret/tiller-secret"
[ℹ]  created "flux:Secret/flux-helm-tls-cert"
[ℹ]  created "flux:Secret/tiller-secret"
[!]  Note: certificate secrets aren't added to the Git repository for security reasons
[ℹ]  Applying manifests
[ℹ]  deleted "flux:Deployment.apps/flux"
[ℹ]  deleted "CustomResourceDefinition.apiextensions.k8s.io/helmreleases.helm.fluxcd.io"
[ℹ]  deleted "flux:Deployment.apps/flux-helm-operator"
[ℹ]  deleted "flux:ServiceAccount/flux-helm-operator"
[ℹ]  deleted "ClusterRole.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  deleted "ClusterRoleBinding.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  deleted "flux:ServiceAccount/tiller"
[ℹ]  deleted "ClusterRoleBinding.rbac.authorization.k8s.io/tiller"
[ℹ]  deleted "flux:ServiceAccount/helm"
[ℹ]  deleted "flux:Role.rbac.authorization.k8s.io/tiller-user"
[ℹ]  deleted "kube-system:RoleBinding.rbac.authorization.k8s.io/tiller-user-binding"
[ℹ]  deleted "flux:Service/memcached"
[ℹ]  deleted "flux:ServiceAccount/flux"
[ℹ]  deleted "ClusterRole.rbac.authorization.k8s.io/flux"
[ℹ]  deleted "ClusterRoleBinding.rbac.authorization.k8s.io/flux"
[ℹ]  deleted "flux:Deployment.extensions/tiller-deploy"
[ℹ]  deleted "flux:Service/tiller-deploy"
[ℹ]  deleted "flux:Deployment.apps/memcached"
[ℹ]  deleted "flux:ConfigMap/flux-helm-tls-ca-config"
[ℹ]  created "flux:Deployment.apps/flux"
[ℹ]  created "CustomResourceDefinition.apiextensions.k8s.io/helmreleases.helm.fluxcd.io"
[ℹ]  created "flux:Deployment.apps/flux-helm-operator"
[ℹ]  created "flux:ServiceAccount/flux-helm-operator"
[ℹ]  created "ClusterRole.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  created "flux:ServiceAccount/tiller"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/tiller"
[ℹ]  created "flux:ServiceAccount/helm"
[ℹ]  created "flux:Role.rbac.authorization.k8s.io/tiller-user"
[ℹ]  created "kube-system:RoleBinding.rbac.authorization.k8s.io/tiller-user-binding"
[ℹ]  created "flux:Service/memcached"
[ℹ]  created "flux:ServiceAccount/flux"
[ℹ]  created "ClusterRole.rbac.authorization.k8s.io/flux"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/flux"
[ℹ]  created "flux:Deployment.extensions/tiller-deploy"
[ℹ]  created "flux:Service/tiller-deploy"
[ℹ]  created "flux:Deployment.apps/memcached"
[ℹ]  created "flux:ConfigMap/flux-helm-tls-ca-config"
[ℹ]  Waiting for Helm Operator to start
ERROR: logging before flag.Parse: E0829 17:42:37.502328   94768 portforward.go:331] an error occurred forwarding 49808 -> 3030: error forwarding port 3030 to pod 707a9ee20bea9eba294bb18eabb91f3802493b94d61e7ac10d62a65d20a8f9d6, uid : exit status 1: 2019/08/29 08:42:37 socat[22727] E connect(5, AF=2 127.0.0.1:3030, 16): Connection refused
[!]  Helm Operator is not ready yet (Get http://127.0.0.1:49808/healthz: EOF), retrying ...
[...]
[ℹ]  Helm Operator started successfully
[ℹ]  see https://docs.fluxcd.io/projects/helm-operator for details on how to use the Helm Operator
[ℹ]  Waiting for Flux to start
[ℹ]  Flux started successfully
[ℹ]  see https://docs.fluxcd.io/projects/flux for details on how to use Flux
[ℹ]  Committing and pushing manifests to git@github.com:marccarre/my-gitops-repo.git
[master 56fb8c9] Add Initial Flux configuration
 1 file changed, 38 insertions(+), 38 deletions(-)
 rewrite flux/tiller-ca-cert-configmap.yaml (83%)
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 8 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 2.14 KiB | 2.14 MiB/s, done.
Total 4 (delta 1), reused 0 (delta 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To github.com:marccarre/my-gitops-repo.git
   a14f424..56fb8c9  master -> master
[ℹ]  Flux will only operate properly once it has write-access to the Git repository
[ℹ]  please configure git@github.com:marccarre/my-gitops-repo.git so that the following Flux SSH public key has write access to it
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC6wyI1Zzc7bFJgpWbrTA6xISdPeavT9zLdbMGf/X9NBmJt+wj1TG0SOGkMSpS670KoNdWu8KdSh77uPkif9MDn2f/oaOdfMwq6nOAQoPdTHpmdx6friliLkoV0RozFkuLPWfg3xXQIkPanRX6APUivOXoqI4jRXt9Ghj62BZkYDNIzzkQ5dOwbZB+nv4+q9X8fC4okMlU68QWosan4qb1mDB0LwIYyhqTAqwTd2aWww+/e0jy/vEdWdD1LVTVb4S8DDcXiijFa32nbVmU40Xm5xBH9+1jNLo/REM7PjuVvah1XFmKpyPZhYglPe/bJsHH+A9P2pY9OEsq/4dzCTQFx
```

### CODE REVIEW

**Comment by errordeveloper** [comment]
@marccarre thanks for getting new general methods in `pkg/kubernetes`, as I mentioned earlier, this package should ideally find a home elsewhere (as it's something that is currently missing from `k8s.io/cli-runtime` and `k8s.io/client-go`). Once you are done with general review items, it would be great if you could add some tests for the new methods, it should be relatively straight-forward as there are some tests for object creation already.

File: pkg/gitops/flux/installer.go [hunk 1]
```diff
@@ -302,6 +319,8 @@ func (fi *Installer) applySecrets(secrets []*corev1.Secret) error {
 		}
 		secretMap[id] = secretBytes
 	}
+	// Add the flux manifest, in case it needs to be created first:
+	secretMap[fluxNamespaceFileName] = manifests[fluxNamespaceFileName]
```
Comment (2opremio) [inline pkg/gitops/flux/installer.go]: This is a side-effect not made clear by the function's name. Can we ensure the namespace exists outside this function?

File: pkg/kubernetes/client.go [hunk 2]
```diff
@@ -224,6 +225,62 @@ func (c *RawClient) createOrReplaceObject(object runtime.RawExtension, plan bool
...
+	if status != "" {
+		logger.Info(status)
+	}
+	return nil
+}
+
+// Exists checks if the Kubernetes resources in the provided manifest exist or
+// not, and returns a map[<namespace>]map[<name>]bool to indicate each
+// resource's existence.
+func (c *RawClient) Exists(manifest []byte) (map[string]map[string]bool, error) {
```
Comment (2opremio) [inline pkg/kubernetes/client.go]: I think this function does too much. Can we do the parsing and existence checks separately?

Also, do we really need to check for existence of each element? Can't we just try to create all the resources in a manifest and do the delete-create on a case-by-case basis? 

By doing that we can piggyback on the manifest parsing done by `CreateOrReplace`, right now we are parsing the manifests twice.

Alternatively we can do the parsing ourselves and make `CreateOrReplace` operate on RawResources.

Comment (2opremio) [reply pkg/kubernetes/client.go]: The more I think about it, the better it feels for `Exists()` and `CreateOrReplace()` to operate on resources and not on full manifests:

1. We decouple parsing and operations (which makes the code easier to read)
2. It simplifies handling of individual resources as opposed to managing a chunk of resources living in a manifest.

File: pkg/kubernetes/client.go [hunk 3]
```diff
@@ -360,3 +409,66 @@ func (r *RawResource) CreatePatchOrReplace() error {
...
+	if !exists {
+		return "", nil
+	}
+	propagationPolicy := metav1.DeletePropagationForeground
+	if _, err := r.Helper.DeleteWithOptions(r.Info.Namespace, r.Info.Name, &metav1.DeleteOptions{
+		PropagationPolicy: &propagationPolicy,
+	}); err != nil {
+		return "", err
+	}
+	if err := r.waitForDeletion(); err != nil {
```
Comment (2opremio) [inline pkg/kubernetes/client.go]: Shouldn't we make synchronous operations optional? Otherwise it would be more readable to call the function `DeleteSynchronously()`

File: pkg/kubernetes/client.go (hunk 2 above)
Comment (2opremio) [reply pkg/kubernetes/client.go]: Oh, Actually I saw you already create `Exists()` for resources. That should make it easier

File: pkg/gitops/flux/installer.go (hunk 1 above)
Comment (marccarre) [reply pkg/gitops/flux/installer.go]: Sure. Will fix. 👍 
(I was tempted to do it in first place as well, but wanted to get something out first, so didn't take the time initially.)

File: pkg/kubernetes/client.go (hunk 3 above)
Comment (marccarre) [reply pkg/kubernetes/client.go]: I also considered that, but since we don't need it now... 🙂 
I'll rename then. We can refine when we actually need a `Delete` which is asynchronous.

Comment (errordeveloper) [reply pkg/kubernetes/client.go]: Yes, it does look like we should make it optional or rename the function. If you chose to rename, please use `DeleteSync`, as that follows the naming used in `pkg/cfn/manager` (e.g. `tasks.DoAllSync` and `DeleteStackBySpecSync`).

File: pkg/kubernetes/client.go (hunk 2 above)
Comment (errordeveloper) [reply pkg/kubernetes/client.go]: How about `CheckAllResourcesExist(manifest []byte) (bool, error)`? That would simplify things and is rather sufficient for current usage.

Comment (marccarre) [reply pkg/kubernetes/client.go]: Having the parsing and actual operations done separately would mean changing the API of `RawClient` so that `CreateOrReplace`, `Exists` and `Delete` take `[]runtime.RawExtension` in, instead of `[]byte` like it does now, hence exposing the implementation details a bit more.
Do we really want that? 🤔 

(I personally don't mind parsing twice for the sake of, arguably, keeping this a bit simpler, at the cost of redundant work and associated CPU and I/O waste, given this is still pretty light in absolute. But I also don't really like the abstractions we have now. So I'm not against refactoring this either, but maybe just not now... WDYT? ⌚️)

> Can't we just try to create all the resources in a manifest and do the delete-create on a case-by-case basis?

Sorry, but not sure I understand what you have in mind there? 😅

Comment (marccarre) [reply pkg/kubernetes/client.go]: `CheckAllResourcesExist` isn't really what I need here though. I only need to check if only field in one `Secret` in one file is present or not. Are we 100% sure nothing adds any other field to that `Secret`?

File: pkg/kubernetes/client.go (hunk 3 above)
Comment (marccarre) [reply pkg/kubernetes/client.go]: Damn GitHub, your reply didn't show up @errordeveloper, renaming again.

**Comment by marccarre** [comment]
### Manual testing (after all changes)

#### First installation

```
[ℹ]  Writing Flux manifests
[ℹ]  created "Namespace/flux"
[ℹ]  Applying Helm TLS Secret(s)
[ℹ]  created "flux:Secret/flux-helm-tls-cert"
[ℹ]  created "flux:Secret/tiller-secret"
[!]  Note: certificate secrets aren't added to the Git repository for security reasons
[ℹ]  Applying manifests
[ℹ]  created "flux:ConfigMap/flux-helm-tls-ca-config"
[ℹ]  created "flux:Deployment.apps/flux"
[ℹ]  created "flux:ServiceAccount/tiller"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/tiller"
[ℹ]  created "flux:ServiceAccount/helm"
[ℹ]  created "flux:Role.rbac.authorization.k8s.io/tiller-user"
[ℹ]  created "kube-system:RoleBinding.rbac.authorization.k8s.io/tiller-user-binding"
[ℹ]  created "flux:Service/tiller-deploy"
[ℹ]  created "flux:Deployment.apps/memcached"
[ℹ]  created "flux:Deployment.apps/flux-helm-operator"
[ℹ]  created "flux:Secret/flux-git-deploy"
[ℹ]  created "flux:Service/memcached"
[ℹ]  created "flux:ServiceAccount/flux-helm-operator"
[ℹ]  created "ClusterRole.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  created "CustomResourceDefinition.apiextensions.k8s.io/helmreleases.helm.fluxcd.io"
[ℹ]  created "flux:Deployment.extensions/tiller-deploy"
[ℹ]  created "flux:ServiceAccount/flux"
[ℹ]  created "ClusterRole.rbac.authorization.k8s.io/flux"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/flux"
[ℹ]  Waiting for Helm Operator to start
[...]
[ℹ]  Flux will only operate properly once it has write-access to the Git repository
[ℹ]  please configure git@github.com:marccarre/my-gitops-repo.git so that the following Flux SSH public key has write access to it
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCyzY6zP4fNjSG9zmPN8gNVO+VOGiBUGE2Ajv7JaAF2Vg1LRRfraNGjiUMCAL5DVw0qyVjR3y2S/msunnD268tHaEkIfuIbkVB0uThR37w9xflPjb+Pa9RCQtJtgGeJBMC4P/RX4BQBpHnCouJODJSJ46VKD3Xy/zTFcRf6rSbGff8fFWRnJdn8/vqc21nWWe6hnWvWM3hFiInbrQv/ULZKaJDyBjnZ8gRL4GRq1Lt46EYhFf3EgJvX+qx3irK59Xk3+WFrBO8e9gIlrIWEX0C9xZucuqUQZoaAT/LxLoe+d4z+mlGM2VbFc2e6HidHOIIxA7dYLapr3FqISveuNf0V
```

#### Second installation

```
[ℹ]  Writing Flux manifests
[ℹ]  Applying Helm TLS Secret(s)
[ℹ]  deleted "flux:Secret/tiller-secret"
[ℹ]  deleted "flux:Secret/flux-helm-tls-cert"
[ℹ]  created "flux:Secret/tiller-secret"
[ℹ]  created "flux:Secret/flux-helm-tls-cert"
[!]  Note: certificate secrets aren't added to the Git repository for security reasons
[ℹ]  Applying manifests
[ℹ]  deleted "flux:Service/tiller-deploy"
[ℹ]  deleted "flux:ServiceAccount/tiller"
[ℹ]  deleted "ClusterRoleBinding.rbac.authorization.k8s.io/tiller"
[ℹ]  deleted "flux:ServiceAccount/helm"
[ℹ]  deleted "flux:Role.rbac.authorization.k8s.io/tiller-user"
[ℹ]  deleted "kube-system:RoleBinding.rbac.authorization.k8s.io/tiller-user-binding"
[ℹ]  deleted "flux:Deployment.extensions/tiller-deploy"
[ℹ]  deleted "flux:Deployment.apps/flux"
[ℹ]  deleted "flux:ServiceAccount/flux-helm-operator"
[ℹ]  deleted "ClusterRole.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  deleted "ClusterRoleBinding.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  deleted "flux:ConfigMap/flux-helm-tls-ca-config"
[ℹ]  deleted "flux:Service/memcached"
[ℹ]  deleted "CustomResourceDefinition.apiextensions.k8s.io/helmreleases.helm.fluxcd.io"
[ℹ]  deleted "flux:Deployment.apps/flux-helm-operator"
[ℹ]  deleted "flux:Deployment.apps/memcached"
[ℹ]  deleted "flux:ServiceAccount/flux"
[ℹ]  deleted "ClusterRole.rbac.authorization.k8s.io/flux"
[ℹ]  deleted "ClusterRoleBinding.rbac.authorization.k8s.io/flux"
[ℹ]  created "flux:Service/tiller-deploy"
[ℹ]  created "flux:ServiceAccount/tiller"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/tiller"
[ℹ]  created "flux:ServiceAccount/helm"
[ℹ]  created "flux:Role.rbac.authorization.k8s.io/tiller-user"
[ℹ]  created "kube-system:RoleBinding.rbac.authorization.k8s.io/tiller-user-binding"
[ℹ]  created "flux:Deployment.extensions/tiller-deploy"
[ℹ]  created "flux:Deployment.apps/flux"
[ℹ]  created "flux:ServiceAccount/flux-helm-operator"
[ℹ]  created "ClusterRole.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  created "flux:ConfigMap/flux-helm-tls-ca-config"
[ℹ]  created "flux:Service/memcached"
[ℹ]  created "CustomResourceDefinition.apiextensions.k8s.io/helmreleases.helm.fluxcd.io"
[ℹ]  created "flux:Deployment.apps/flux-helm-operator"
[ℹ]  created "flux:Deployment.apps/memcached"
[ℹ]  created "flux:ServiceAccount/flux"
[ℹ]  created "ClusterRole.rbac.authorization.k8s.io/flux"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/flux"
[ℹ]  Waiting for Helm Operator to start
[...]
[ℹ]  Flux will only operate properly once it has write-access to the Git repository
[ℹ]  please configure git@github.com:marccarre/my-gitops-repo.git so that the following Flux SSH public key has write access to it
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCyzY6zP4fNjSG9zmPN8gNVO+VOGiBUGE2Ajv7JaAF2Vg1LRRfraNGjiUMCAL5DVw0qyVjR3y2S/msunnD268tHaEkIfuIbkVB0uThR37w9xflPjb+Pa9RCQtJtgGeJBMC4P/RX4BQBpHnCouJODJSJ46VKD3Xy/zTFcRf6rSbGff8fFWRnJdn8/vqc21nWWe6hnWvWM3hFiInbrQv/ULZKaJDyBjnZ8gRL4GRq1Lt46EYhFf3EgJvX+qx3irK59Xk3+WFrBO8e9gIlrIWEX0C9xZucuqUQZoaAT/LxLoe+d4z+mlGM2VbFc2e6HidHOIIxA7dYLapr3FqISveuNf0V
```

File: pkg/kubernetes/client.go (hunk 2 above)
Comment (errordeveloper) [reply pkg/kubernetes/client.go]: Given this is all about the fact that Flux manifests contain an empty secret, and going forward we have https://github.com/weaveworks/eksctl/issues/1213 to close on, I'd leave this as an inevitable consequence of having to handle the secret in a special way.

Comment (marccarre) [reply pkg/kubernetes/client.go]: Merged. We need to move forward. I'm happy to refine, refactor, clean-up, etc. later once clear how we want to proceed. 👍
